### PR TITLE
blitz-325-use-current-values-in-db-in-transformations

### PR DESCRIPTION
**Pull request by samuel-blitz**

### CODE REVIEW

File: src/stateMachine/mutation/BQL/preQuery.ts [hunk 1]
```diff
@@ -59,6 +60,7 @@ export const mutationPreQuery = async (
 
 	// 2. Check operations to make sure they include: Delete, Unlink, Link, Replace, Updater
 	const ops: string[] = [];
+	let has$Fields = false;
```
Comment (lveillard) [inline src/stateMachine/mutation/BQL/preQuery.ts:63]: This we will need to find a more immutable way to do it 🤔

File: tests/mocks/schema.ts [hunk 2]
```diff
@@ -481,6 +666,44 @@ export const schema: BormSchema = {
...
+											things: dbThings.map((id: string) => {
+												return {
+													$op: 'delete',
+													$id: id,
+													$fields: ['cascadeRelations'],
+												};
+											}),
+										};
+									} else {
+										return {};
```
Comment (aguspdana) [inline tests/mocks/schema.ts:698]: We don't need an else block after "if return"

File: tests/mocks/schema.ts [hunk 3]
```diff
@@ -671,6 +894,54 @@ export const schema: BormSchema = {
...
+											if (dbExpression || dbValues) {
+												throw new Error('The df should have no expression and no values');
+											}
+											return true;
+										} else {
+											return true;
+										}
+									} else {
+										return true;
+									}
```
Comment (aguspdana) [inline tests/mocks/schema.ts:938]: We can flatten this nested if/else

File: tests/mocks/schema.ts [hunk 4]
```diff
@@ -693,14 +964,82 @@ export const schema: BormSchema = {
...
+										}
+										if ($id === 'mf2-expression-2') {
+											if (dbId !== 'mf2-expression-2') {
+												throw new Error('The df should have one expression');
+											}
+										}
+										return true;
+									} else {
+										return true;
+									}
```
Comment (aguspdana) [inline tests/mocks/schema.ts:997]: We can flatten this nested if/else

File: tests/mocks/schema.ts [hunk 5]
```diff
@@ -693,14 +964,82 @@ export const schema: BormSchema = {
...
+										}
+										if ($id === 'mf2-dataValue-2') {
+											if (dbId !== 'mf2-dataValue-2') {
+												throw new Error('The df should have one dv');
+											}
+										}
+										return true;
+									} else {
+										return true;
+									}
```
Comment (aguspdana) [inline tests/mocks/schema.ts:1036]: We can flatten this nested if/else

File: src/stateMachine/mutation/BQL/enrich.ts [hunk 6]
```diff
@@ -72,9 +75,13 @@ export const enrichBQLMutation = (
 				if ('$root' in value) {
 					// This is hte $root object, we will split the real root if needed in this iteration
 				} else if (!('$thing' in value || '$entity' in value || '$relation' in value)) {
+					const toIgnore = ['$fields', '$dbNode'];
+					const paths: string[] = meta.nodePath?.split('.') || [];
```
Comment (aguspdana) [inline src/stateMachine/mutation/BQL/enrich.ts:79]: We don't need to specify the type here.

File: src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+			}
+		});
+	};
+
+	if (!blocks) {
+		throw new Error('[BQLE-M-PQ-1] No blocks found');
+	}
+	/// 1. Convert the mutation to a query to get all $fields
+	const convertMutationToQuery = (blocks: FilledBQLMutationBlock[]) => {
+		const processBlock = (block: FilledBQLMutationBlock, root?: boolean) => {
```
Comment (aguspdana) [inline src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts:50]: We can move out this function and update the name.

File: src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+					}
+				} else {
+					// @ts-expect-error todo
+					filteredBlock[k] = block[k];
+				}
+			}
+			return {
+				...filteredBlock,
+				$fields,
+			};
```
Comment (aguspdana) [inline src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts:94]: We don't need to create a new object. We can set `$fields` to filteredBlocks` and return it.

File: src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+			const ids = Array.isArray($objectPath.ids) ? `[${$objectPath.ids}]` : $objectPath.ids;
+			const final = `${root}.${ids}___${$objectPath.key}`;
+
+			const new$objectPath = {
+				beforePath: final,
+				ids: idField,
+				key,
+			};
+			return new$objectPath;
+		} else {
```
Comment (aguspdana) [inline src/stateMachine/mutation/BQL/enrichSteps/preHookDependencies.ts:124]: We don't need else block after "if return"

File: src/stateMachine/mutation/BQL/guards/dependenciesGuard.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+				hasFields = true;
+				return;
+			}
+		} else if (isObject(value)) {
+			// @ts-expect-error todo
+			if (value.$fields) {
+				hasFields = true;
+				return;
+			}
+		}
```
Comment (aguspdana) [inline src/stateMachine/mutation/BQL/guards/dependenciesGuard.ts:37]: We can flatten this nested if/else.

File: src/stateMachine/mutation/BQL/guards/preQueryGuard.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	if (config.mutation?.preQuery === false) {
+		return false;
+	}
+
+	const ops: string[] = [];
+
+	traverse(blocks, ({ parent, key, value }) => {
+		if (parent && key && !key.includes('$') && isObject(parent)) {
```
Comment (aguspdana) [inline src/stateMachine/mutation/BQL/guards/preQueryGuard.ts:20]: I guess it should be `!key.startsWith('$')`

**Pull request merged**